## **4** Conclusion
<font color = red>[15 marks]</font> <br>

### **4.1** Final Insights and Recommendations
<font color = red>[15 marks]</font> <br>

Conclude your analyses here. Include all the outcomes you found based on the analysis.

Based on the insights, frame a concluding story explaining suitable parameters such as location, time of the day, day of the week etc. to be kept in mind while devising a strategy to meet customer demand and optimise supply.

**4.1.1** <font color = red>[5 marks]</font> <br>
Recommendations to optimize routing and dispatching based on demand patterns and operational inefficiencies

In [ ]:
print("#### 1. Dynamic Dispatching based on Hourly/Daily Demand")

# Re-calculate hourly trip counts if not readily available from previous cells
hourly_trips = df['pickup_hour'].value_counts().sort_index()

# Identify the top 3 busiest hours overall
top_n = 3
busiest_hours_overall = hourly_trips.nlargest(top_n)
print(f"\nTop {top_n} Busiest Hours Overall (Total Trips):\n{busiest_hours_overall}")

# Identify peak hours for weekdays vs. weekends
print("\nPeak hours for Weekdays vs. Weekends:")
print(hourly_traffic_pattern)

# Suggest a strategy based on peak hours and top pickup zones
# For example, identify top pickup zones during the absolute busiest hour
busiest_hour_val = busiest_hours_overall.index[0] # Get the hour with most trips
top_pickup_zones_busiest_hour = df[df['pickup_hour'] == busiest_hour_val]['pu_location_id'].value_counts().nlargest(5)

print(f"\nTop 5 Pickup Zones during the busiest hour ({busiest_hour_val}:00):\n{top_pickup_zones_busiest_hour}")
print("\nRecommendation: During these peak hours and in these top pickup zones, increase taxi availability through targeted dispatching.")


print("\n#### 2. Addressing Slow Routes (Operational Inefficiencies)")

# Display the top 10 slowest routes identified earlier
print("\nTop 10 slowest routes (miles per hour):\n")
print(slowest_routes.head(10))
print("\nRecommendation: Analyze these routes further for traffic patterns, road conditions, or specific times of congestion. Provide drivers with real-time traffic updates and suggest alternative routes, or adjust estimated travel times for passengers during these periods.")

#### 1. Dynamic Dispatching based on Hourly/Daily Demand

Top 3 Busiest Hours Overall (Total Trips):
pickup_hour
18    10449
17    10182
15     9602
Name: count, dtype: int64

Peak hours for Weekdays vs. Weekends:
pickup_hour    0     1     2    3    4    5     6     7     8     9   ...  \
day_type                                                              ...   
Weekday      1768   873   476  313  305  606  1716  3595  4769  5087  ...   
Weekend      2306  1972  1526  977  517  226   325   523   777  1226  ...   

pickup_hour    14    15    16    17    18    19    20    21    22    23  
day_type                                                                 
Weekday      6755  6957  6841  7550  7884  6947  6030  5836  5193  3770  
Weekend      2615  2645  2705  2632  2565  2398  2037  2007  1986  1782  

[2 rows x 24 columns]

Top 5 Pickup Zones during the busiest hour (18:00):
pu_location_id
161    596
237    546
236    485
132    446
162    432
Name: count, dtype: int64

Recommen

**4.1.2** <font color = red>[5 marks]</font> <br>

Suggestions on strategically positioning cabs across different zones to make best use of insights uncovered by analysing trip trends across time, days and months.

In [ ]:
print("#### 3. Strategic Cab Positioning (Resource Allocation)")

print("\nTop 10 zones with the highest pickup-to-dropoff ratios (more pickups than dropoffs, potential for empty cabs to accumulate):")
# Re-display top 10 high ratio zones
print(top_10_high_ratio_zones.merge(zones[['LocationID', 'zone', 'borough']], on='LocationID', how='left'))

print("\nTop 10 zones with the lowest pickup-to-dropoff ratios (more dropoffs than pickups, potential for cabs to be needed):")
# Re-display top 10 low ratio zones
print(top_10_low_ratio_zones.merge(zones[['LocationID', 'zone', 'borough']], on='LocationID', how='left'))

print("\nBusiest hours for Weekdays vs. Weekends:")
print(hourly_traffic_pattern)

print("\nTop 5 Busiest Pickup Zones during peak demand hours (e.g., 6 PM):")
# (Assuming busiest_hour_val is still 18 from previous execution)
print(top_pickup_zones_busiest_hour.rename('Trips').reset_index().merge(zones[['LocationID', 'zone', 'borough']], left_on='pu_location_id', right_on='LocationID', how='left'))

print("\nRecommendation: Utilize zones with high pickup-to-dropoff ratios (e.g., airports, major business districts during peak times) as primary dispatch points. Conversely, monitor zones with low pickup-to-dropoff ratios and direct available cabs to these areas from nearby high-demand pickup zones to ensure supply meets demand. Implement dynamic re-positioning based on real-time and historical hourly/daily patterns.")

#### 3. Strategic Cab Positioning (Resource Allocation)

Top 10 zones with the highest pickup-to-dropoff ratios (more pickups than dropoffs, potential for empty cabs to accumulate):
   LocationID  total_pickups  total_dropoffs  pickup_to_dropoff_ratio  \
0         253            1.0             0.0                      inf   
1         187            1.0             0.0                      inf   
2         204            1.0             0.0                      inf   
3          70          631.0            81.0                 7.790123   
4         132         7830.0          1510.0                 5.185430   
5         138         4352.0          1547.0                 2.813187   
6         264         2002.0          1115.0                 1.795516   
7         186         5344.0          3303.0                 1.617923   
8           6            6.0             4.0                 1.500000   
9         214            3.0             2.0                 1.500000   

              

**4.1.3** <font color = red>[5 marks]</font> <br>
Propose data-driven adjustments to the pricing strategy to maximize revenue while maintaining competitive rates with other vendors.

#### 1. Dynamic Pricing based on Demand and Time

This recommendation is supported by the hourly and daily demand patterns and the average fare per mile by hour and day of the week. Higher demand and higher average fares per mile during certain periods suggest opportunities for dynamic pricing.

In [ ]:
print("Hourly Traffic Patterns (Weekday vs. Weekend):")
display(hourly_traffic_pattern)

print("\nAverage Fare Per Mile by Hour of Day:")
display(avg_fare_per_mile_hourly)

print("\nAverage Fare Per Mile by Day of Week:")
display(avg_fare_per_mile_daily)

# Re-plotting for visual context if needed, otherwise refer to previous plots

Hourly Traffic Patterns (Weekday vs. Weekend):


pickup_hour,0,1,2,3,4,5,6,7,8,9,...,14,15,16,17,18,19,20,21,22,23
day_type,,,,,,,,,,,,,,,,,,,,,
Weekday,1768,873,476,313,305,606,1716,3595,4769,5087,...,6755,6957,6841,7550,7884,6947,6030,5836,5193,3770
Weekend,2306,1972,1526,977,517,226,325,523,777,1226,...,2615,2645,2705,2632,2565,2398,2037,2007,1986,1782



Average Fare Per Mile by Hour of Day:


,pickup_hour,fare_per_mile
0,0,12.144519
1,1,18.954546
2,2,14.598594
3,3,12.956825
4,4,11.190351
5,5,15.526000
6,6,10.592745
7,7,11.994476
8,8,15.860384
9,9,15.037289



Average Fare Per Mile by Day of Week:


,pickup_day_of_week,fare_per_mile
0,Monday,15.679810
1,Tuesday,16.485704
2,Wednesday,15.901510
3,Thursday,16.035341
4,Friday,15.398346
5,Saturday,14.567089
6,Sunday,16.774366


#### 2. Tiered Pricing Review for Competitiveness

This strategy is informed by comparing the average fare per mile across different distance tiers for each vendor. This helps identify where pricing might be out of sync with competitors or market expectations.

In [ ]:
print("Average Fare Per Mile by Vendor and Distance Tier:")
display(avg_fare_per_mile_tiered)

Average Fare Per Mile by Vendor and Distance Tier:


,vendor_id,distance_tier,fare_per_mile
0,1,up_to_2_miles,16.409844
1,1,2_to_5_miles,8.739552
2,1,more_than_5_miles,5.805477
3,2,up_to_2_miles,24.575189
4,2,2_to_5_miles,9.052282
5,2,more_than_5_miles,5.961657


#### 3. Optimize Passenger Count Pricing

Understanding the average fare per mile per passenger helps in evaluating if the current pricing adequately accounts for varying group sizes, especially since the base fare isn't strongly correlated with passenger count.

In [ ]:
print("Average Fare Per Mile Per Passenger for different Passenger Counts:")
display(average_fare_per_mile_per_passenger)

Average Fare Per Mile Per Passenger for different Passenger Counts:


,passenger_count,fare_per_mile_per_passenger
0,1.0,15.670481
1,2.0,8.652352
2,3.0,4.458309
3,4.0,5.001194
4,5.0,2.566405
5,6.0,2.113981


#### 4. Incentivize Tipping through Digital Payments

The analysis of tipping behavior, particularly the stark difference in payment types for high vs. low tip percentages, supports promoting digital payment methods.

In [ ]:
print("--- Payment Type Distribution for Low Tip Trips (<10%) ---")
# Ensure payment_type_map is accessible for re-evaluation if needed
print(low_tip_trips['payment_type'].value_counts(normalize=True).map(payment_type_map))

print("\n--- Payment Type Distribution for High Tip Trips (>25%) ---")
print(high_tip_trips['payment_type'].value_counts(normalize=True).map(payment_type_map))

--- Payment Type Distribution for Low Tip Trips (<10%) ---
payment_type
2    NaN
1    NaN
4    NaN
3    NaN
Name: proportion, dtype: object

--- Payment Type Distribution for High Tip Trips (>25%) ---
payment_type
1    Credit card
Name: proportion, dtype: object


#### 5. Surcharge Optimization

Reviewing the frequency of surcharge application helps understand their impact on total fare perception and ensures they are applied as intended.

In [ ]:
print("Surcharge Application Frequency (Percentage of Trips):")
display(surcharge_frequency_series)

Surcharge Application Frequency (Percentage of Trips):


,0
extra,61.183103
mta_tax,99.240945
improvement_surcharge,99.967671
congestion_surcharge,92.414160
airport_fee,8.724086


#### 4. Pricing Strategy Adjustments

Based on the analysis, here are data-driven adjustments to the pricing strategy:

1.  **Dynamic Pricing based on Demand and Time:**
    *   **Peak Hours/Days:** Implement surge pricing during identified peak hours (e.g., weekday evenings 5-7 PM, weekend late nights) and busiest days (weekends, especially Saturday). The `avg_fare_per_mile_hourly` and `avg_fare_per_mile_daily` data show fluctuations that can be capitalized on.
    *   **Off-Peak Promotions:** Consider offering discounts or loyalty incentives during off-peak hours (e.g., early mornings 4-6 AM, Sunday afternoons) to stimulate demand and utilize idle capacity, as indicated by lower pickup counts and tip percentages in those periods.

2.  **Tiered Pricing Review for Competitiveness:**
    *   **Short Trips (up to 2 miles):** The `avg_fare_per_mile_tiered` data indicates Vendor 2 has a significantly higher average fare per mile for short trips (up to 2 miles) compared to Vendor 1. Review this pricing tier to ensure it remains competitive while still covering operational costs. If Vendor 1's rates are more appealing for short distances, it might be losing customers there.
    *   **Longer Trips (more than 5 miles):** Both vendors have similar, lower fare-per-mile rates for longer trips. This suggests a more standardized market for these distances.

3.  **Optimize Passenger Count Pricing:**
    *   While `passenger_count` has a very weak correlation with `fare_amount`, the `average_fare_per_mile_per_passenger` table shows that the effective cost per passenger significantly decreases with more passengers. Currently, fares are not primarily passenger-count dependent. Consider if a slight increase in fare (e.g., a small per-person surcharge beyond 2-3 passengers for certain trip types) could be implemented without deterring groups, especially if larger vehicles are dispatched for higher counts. This could be tested in conjunction with the customer experience insights.

4.  **Incentivize Tipping through Digital Payments:**
    *   The analysis on tip percentages highlighted that high tip percentages are exclusively associated with credit card payments. Promote digital payment methods (credit card in-app payments) more heavily, as this directly correlates with higher tip percentages and likely total driver earnings. This could be through preferred payment options or even small incentives for drivers for digital payments.

5.  **Surcharge Optimization:**
    *   Regularly review the impact of frequently applied surcharges like 'improvement_surcharge', 'mta_tax', and 'congestion_surcharge' (all applied in over 90% of trips). While these are often fixed, understanding their impact on total fare perception is crucial. The 'airport_fee' is applied less frequently (around 8.7%), suggesting its application is situational, which is appropriate.

By strategically adjusting pricing based on these factors, the taxi operation can enhance revenue, optimize fleet utilization, and maintain a competitive edge in the market.